src/upgrade.py

```diff
@@ -0,0 +1,219 @@
+        """Return the max ordinal."""
+        return len(self.peer_relation.units)
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        """Run pre-upgrade checks."""
+        fail_message = "Pre-upgrade check failed. Cannot upgrade."
+
+        def _online_instances(status_dict: dict) -> int:
+            """Return the number of online instances from status dict."""
+            return [
+                item["status"] for item in status_dict["defaultreplicaset"]["topology"].values()
+            ].count("online")
+
+        if cluster_status := self.charm._mysql.get_cluster_status():
+            if _online_instances(cluster_status) < self.highest_ordinal + 1:
+                # case any offline unit is found
+                raise ClusterNotReadyError(
+                    message=fail_message,
+                    cause="Not all units are online",
+                    resolution="Ensure all units are online in the cluster",
+                )
+        else:
+            raise ClusterNotReadyError(
+                message=fail_message,
+                cause="Failed to retrieve cluster status",
+                resolution="Check cluster health from cluster status",
+            )
+
+        try:
+            self._pre_upgrade_prepare()
+        except MySQLSetClusterPrimaryError:
+            raise ClusterNotReadyError(
+                message=fail_message,
+                cause="Failed to set primary",
+                resolution="Check the cluster status",
+            )
+        except KubernetesClientError:
+            raise ClusterNotReadyError(
+                message=fail_message,
+                cause="Failed to patch statefulset",
+                resolution="Check kubernetes access policy",
+            )
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        """Log rollback instructions."""
+        logger.info(
+            "Run `juju refresh --revision <previous-revision> mysql-k8s` to initiate the rollback"
+        )
+        logger.info("and `juju run-action mysql-k8s/leader resume-upgrade` to resume the rollback")
+
+    def _pre_upgrade_prepare(self) -> None:
+        """Pre upgrade routine for MySQL.
+
+        Set primary to the first unit to avoid switchover during upgrade and
+        patch statefulset `spec.updateStrategy.rollingUpdate.partition` to the last unit.
+        """
+        if self.charm._mysql.get_primary_label() != f"{self.charm.app.name}-0":
+            # set the primary to the first unit for switchover mitigation
+            self.charm._mysql.set_cluster_primary(
+                self.charm._get_unit_fqdn(f"{self.charm.app.name}/0")
+            )
+
+        self.charm.k8s_helpers.set_rolling_update_partition(partition=self.highest_ordinal)
+
+    def _on_resume_upgrade(self, event: ActionEvent) -> None:
+        """Handle resume upgrade action.
+
+        Continue the upgrade by setting the partition to the next unit.
+        """
```

Comment:
JFYI, after this PR was created last week I generalised this action handler (and also the on for `_on_upgrade_finished`) and added it to https://github.com/canonical/data-platform-libs/pull/77.